source/FormationPositioner.cpp

```diff
@@ -0,0 +1,276 @@
+		if(ship)
+			itCoor = shipPositions.find(ship.get());
+
+		// If the ship is not in the overall table or if it was not
+		// active since the last iteration, then we also remove it.
+		removeShip |= itCoor == shipPositions.end() ||
+				itCoor->second.second != tickTock;
+
+		// Perform removes if we need to.
+		if(removeShip)
+		{
+			// Remove the ship from the ring.
+			Remove(shipIndex);
+			// And remove the ship from the shipsPositions map.
+			if(itCoor != shipPositions.end())
+				shipPositions.erase(itCoor);
+		}
+		else
+		{
+			// Set scaling for next round based on the sizes of the
+			// participating ships.
+			Tally(*ship);
+
+			// Calculate the new coordinate for the current ship.
+			Point &shipRelPos = itCoor->second.first;
+			shipRelPos = *itPos;
+			if(flippedY)
+				shipRelPos.Set(-shipRelPos.X(), shipRelPos.Y());
+			if(flippedX)
+				shipRelPos.Set(shipRelPos.X(), -shipRelPos.Y());
+			++itPos;
+			++shipIndex;
+		}
+	}
+
+	// Switch marker to detect stale/missing ships in the next iteration.
+	tickTock = !tickTock;
+}
+
+
+
+void FormationPositioner::Tally(const Body &body)
+{
+	maxDiameter = max(maxDiameter, body.Radius() * 2.);
+	maxHeight = max(maxHeight, body.Height());
+	maxWidth = max(maxWidth, body.Width());
+}
+
+
+
+void FormationPositioner::CalculateDirection()
+{
+	// Calculate new direction, if the formationLead is moving, then we use the movement vector.
+	// Otherwise we use the facing vector.
+	Point velocity = formationLead->Velocity();
+	Angle desiredDir = velocity.Length() > .1 ? Angle(velocity) : formationLead->Facing();
+
+	Angle deltaDir = desiredDir - direction;
+
+	// Change the desired direction according to rotational settings if that fits better.
+	double symRot = pattern->Rotatable();
+	if(symRot > 0 && fabs(deltaDir.Degrees()) > symRot / 2)
+	{
+		if(deltaDir.Degrees() > 0)
+			symRot = -symRot;
+
+		while(fabs(deltaDir.Degrees() + symRot) < fabs(deltaDir.Degrees()))
+		{
+			desiredDir += Angle(symRot);
+			deltaDir = desiredDir - direction;
+		}
+	}
+
+	// Angle at which to perform longitudinal or transverse mirror instead of turn.
+	constexpr double MIN_FLIP_TRIGGER = 135.;
+
+	// If we are beyond the triggers for flipping, then immediately go to the desired direction.
+	if(fabs(deltaDir.Degrees()) >= MIN_FLIP_TRIGGER &&
+		(pattern->FlippableY() || pattern->FlippableX()))
+	{
+		direction = desiredDir;
+		deltaDir = Angle(0.);
+		if(pattern->FlippableY())
+		{
+			flippedY = !flippedY;
+			positionsTimer = 0;
+		}
+		if(pattern->FlippableX())
+		{
+			flippedX = !flippedX;
+			positionsTimer = 0;
+		}
+	}
+	else
+	{
+		// Turn max 1/4th degree per frame. The game runs at 60fps, so a turn of 180 degrees will take
+		// about 12 seconds.
+		constexpr double MAX_FORMATION_TURN = .25;
+
+		if(deltaDir.Degrees() > MAX_FORMATION_TURN)
+			deltaDir = Angle(MAX_FORMATION_TURN);
+		else if(deltaDir.Degrees() < -MAX_FORMATION_TURN)
+			deltaDir = Angle(-MAX_FORMATION_TURN);
```

Comment:
Might be easier to just use clamp.
```suggestion
		deltaDir = Angle(std::clamp(deltaDir.Degrees(), -MAX_FORMATION_TURN, MAX_FORMATION_TURN))
```